query-planner-js/src/buildQueryPlan.ts

```diff
@@ -0,0 +1,1189 @@
+        fieldDef: TypeNameMetaFieldDef,
+      });
+    }
+
+    const subfields = collectSubfields(context, returnType, fields);
+    splitSubfields(context, fieldPath, subfields, subGroup);
+
+    parentGroup.otherDependentGroups.push(...subGroup.dependentGroups);
+
+    let definition: FragmentDefinitionNode;
+    let selectionSet = selectionSetFromFieldSet(subGroup.fields, returnType);
+
+    if (context.autoFragmentization && subGroup.fields.length > 2) {
+      ({ definition, selectionSet } = getInternalFragment(
+        selectionSet,
+        returnType,
+        context,
+      ));
+      parentGroup.internalFragments.add(definition);
+    }
+
+    // "Hoist" internalFragments of the subGroup into the parentGroup so all
+    // fragments can be included in the final request for the root FetchGroup
+    subGroup.internalFragments.forEach(fragment => {
+      parentGroup.internalFragments.add(fragment);
+    });
+
+    return {
+      scope,
+      fieldNode: {
+        ...fieldNode,
+        selectionSet,
+      },
+      fieldDef,
+    };
+  }
+}
+
+function getInternalFragment(
+  selectionSet: SelectionSetNode,
+  returnType: GraphQLCompositeType,
+  context: QueryPlanningContext
+) {
+  const key = JSON.stringify(selectionSet);
+  if (!context.internalFragments.has(key)) {
+    const name = `__QueryPlanFragment_${context.internalFragmentCount++}__`;
+
+    const definition: FragmentDefinitionNode = {
+      kind: Kind.FRAGMENT_DEFINITION,
+      name: {
+        kind: Kind.NAME,
+        value: name,
+      },
+      typeCondition: {
+        kind: Kind.NAMED_TYPE,
+        name: {
+          kind: Kind.NAME,
+          value: returnType.name,
+        },
+      },
+      selectionSet,
+    };
+
+    const fragmentSelection: SelectionSetNode = {
+      kind: Kind.SELECTION_SET,
+      selections: [
+        {
+          kind: Kind.FRAGMENT_SPREAD,
+          name: {
+            kind: Kind.NAME,
+            value: name,
+          },
+        },
+      ],
+    };
+
+    context.internalFragments.set(key, {
+      name,
+      definition,
+      selectionSet: fragmentSelection,
+    });
+  }
+
+  return context.internalFragments.get(key)!;
+}
+
+function collectFields(
+  context: QueryPlanningContext,
+  scope: Scope<GraphQLCompositeType>,
+  selectionSet: SelectionSetNode,
+  fields: FieldSet = [],
+  visitedFragmentNames: { [fragmentName: string]: boolean } = Object.create(
+    null,
+  ),
+): FieldSet {
+  for (const selection of selectionSet.selections) {
+    switch (selection.kind) {
+      case Kind.FIELD:
+        const fieldDef = context.getFieldDef(scope.parentType, selection);
+        fields.push({ scope, fieldNode: selection, fieldDef });
+        break;
+      case Kind.INLINE_FRAGMENT: {
+        const newScope = context.newScope(getFragmentCondition(selection), scope);
+        if (newScope.possibleTypes.length === 0) {
+          break;
+        }
+
+        newScope.directives = selection.directives;
```

Comment:
Another case of "you described caveats in a commit message but it would be nice to get a FIXME (re the #177 thing)"